src/workerd/io/worker-interface.capnp

```diff
@@ -147,6 +186,277 @@ struct Trace @0x8e8d911203762d34 {
+    # events occuring while an output gate is open, and the output gate fails indicating
+    # that the events are not longer valid, the outcome field will be used to signal.
+    enum SpanOutcome {
+      unknown @0;
+      ok @1;
+      exception @2;
+      canceled @3;
+    }
+
+    outcome @3 :SpanOutcome;
+
+    # The info field is used to provide additional information about the outcome.
+    # These are often dependent on the type of info event that started the span.
+    # This is most typically only used for stage spans.
+    # For instance, a fetch event will have a fetchResponse info field in the outcome.
+    info :union {
+      none @4 :Void;
+      fetch @5 :FetchResponseInfo;
+      actorFlush @6 :ActorFlushInfo;
+      custom @7 :List(Tag);
+    }
+
+    # Any additional metadata specific to the span itself.
+    tags @8 :List(Tag);
+  }
+
+  # A mark is a special event that simply marks a point of interest in the trace.
+  struct Mark {
+    name @0 :Text;
+  }
+
+  # A metric is a special event that represents a metric value of some form injected
+  # into the trace. A metric can be used, for instance, to communicate the current
+  # isolate memory usage at a given point in time.
+  struct Metric {
+    # The key field can be either arbitrary text or a numeric ID. The ID field is used
+    # to identify commonly used metrics defined and injected by the runtime. There is no
+    # enum definition for these here as it is expected the actual definitions will
+    # come from the internal project and not workerd.
+    # Metrics with the same name can appear multiple times, in which case the calculated
+    # logical value is the concatenation of all values into a collection. So, for
+    # instance, if a metric with name "foo" appears with value 1 and then again with
+    # value 2, the logical value of the "foo" metric is [1, 2]. The ordering of such
+    # metrics is significant.
+    # Multiple appearances of the same metric with the same key can be of different
+    # types and units.
+    key :union {
+      text @0 :Text;
+      id @1 :UInt32;
+    }
+    enum Type {
+      # A counter metric, for instance, the number of times a given event has occurred,
+      # the number of requests processed, etc. Counters nearly aways represent a conceptually
+      # monotonically increasing value (conceptually because in reality the value may not
+      # be strictly increasing unit-by-unit due to sampling frequency, etc).
+      counter @0;
+      # A gauge metric, for instance, the current memory heap size, etc. Gauges represent
+      # a snapshot of a value at a given point in time and therefore may increase or decrease.
+      gauge @1;
+    }
+    type @2 :Type;
+    value :union {
+      float64 @3 :Float64;
+      int64 @4 :Int64;
+      uint64 @5 :UInt64;
+    }
+    unit @6 :Text;
+    tags @7 :List(Tag);
+  }
+
+  # The Dropped struct is used to indicate that a trace has dropped a given number of
```

Comment:
This one is more speculative at the current moment and should be considered at risk as I iterate through the implementation. The idea, at least in theory, is to allow the runtime to skip delivery of events if the system becomes too overwhelmed or the event exceeds some threshold. It would serve the same purpose as the recently added `truncated` field -- that is, a specific indication that events have intentionally not been delivered for whatever reason.